pkg/urbit/jets/e/scow.c

```diff
@@ -0,0 +1,385 @@
+                      127, &f,
+                      0)) {
+    return u3m_bail(c3__exit);
+  }
+
+  if (c3n == u3r_sing(f, 0)) {
+    u3_noun f_list = u3qb_flop(f);
+
+    for (u3_noun cur = f_list;
+         _(u3a_is_cell(cur));
+         cur = u3t(cur)) {
+      if (_(u3a_is_cat(u3h(cur)))) {
+        c3_y a, b, c, d;
+        _x_co_four(u3h(cur), &a, &b, &c, &d);
+        out = u3nq('.', a, b, u3nt(c, d, out));
+      } else {
+        // No way to deal with big atoms. fall back.
+        u3z(yod);
+        u3z(out);
+        u3z(f_list);
+        return u3_none;
+      }
+    }
+
+    u3z(f_list);
+    out = u3nc('.', out);
+  }
+
+  // if there isn't a hex list and the h/m/s are all 0, skip printing hours.
+  if (c3n == u3r_sing(f, 0) ||
+      c3n == u3r_sing(hour, 0) ||
+      c3n == u3r_sing(min, 0) ||
+      c3n == u3r_sing(sec, 0)) {
+    if (!_(u3a_is_cat(hour)) ||
+        !_(u3a_is_cat(min)) ||
+        !_(u3a_is_cat(sec))) {
+      // Input is weird, fallback to nock.
+      u3z(yod);
+      u3z(out);
+      return u3_none;
+    }
+
+    c3_y sa, sb, ma, mb, ha, hb;
+    _y_co_two(sec, &sa, &sb);
+    out = u3nq('.', sa, sb, out);
+
+    _y_co_two(min, &ma, &mb);
+    out = u3nq('.', ma, mb, out);
+
+    _y_co_two(hour, &ha, &hb);
+    out = u3nq('.', ha, hb, out);
+
+    out = u3nc('.', out);
+  }
+
+  // We always print the Y.M.D. Unlike others, these numbers are unconstrained
+  // by length, but in practice, the month number and day number can only be up
+  // to two digits because of +yore. We still need to remove 0 prefixes,
+  // though.
+  if (!_(u3a_is_cat(day)) || day > 99 ||
+      !_(u3a_is_cat(month)) || month > 99 ||
+      !_(u3a_is_cat(year))) {
+    // Input is weird, fallback to nock.
+    u3z(yod);
+    u3z(out);
+    return u3_none;
+  }
+
+  c3_y da, db;
+  _y_co_two(day, &da, &db);
+  out = u3nc(db, out);
+  if (da != '0')
+    out = u3nc(da, out);
+  out = u3nc('.', out);
+
+  c3_y ma, mb;
+  _y_co_two(month, &ma, &mb);
+  out = u3nc(mb, out);
+  if (ma != '0')
+    out = u3nc(ma, out);
+  out = u3nc('.', out);
+
+  // suffix the year with a '-' for BC dates
+  if (age == c3n) {
+    out = u3nc('-', out);
+  }
+
+  // The year part is the only place where we have to explicitly loop over the
+  // input because it can be arbitrarily large or small.
+  out = _add_year(year, out);
+
+  out = u3nc('~', out);
+
+  u3z(yod);
+  return out;
+}
+
+static
+u3_noun
+_print_p(u3_atom cor, u3_atom p)
+{
+  // Scramble the raw number to the concealed version.
+  u3_noun hok = u3j_cook("u3we_scow_fein_p", u3k(cor), "fein");
+  u3_atom sxz = u3n_slam_on(hok, u3k(p));
+
+  // Simple galaxy case
+  if (c3y == u3qa_lte(u3k(sxz), 256)) {
+    c3_y a, b, c;
+    po_to_suffix(sxz, &a, &b, &c);
+    u3z(sxz);
+    return u3nq('~', a, b, u3nc(c, 0));
+  }
+
+  u3_atom dyy = u3qc_met(4, u3k(sxz));
+  if (!_(u3a_is_cat(dyy))) {
+    return u3_none;
+  }
+
+  u3_noun list = 0;
+  for (u3_atom imp = 0; imp != dyy; ++imp) {
```

Comment:
I'm concerned about the use of `u3_atom` here. It doesn't look like `imp` is used for anything but separator counting, so it can be of any type. I see now that you `return u3_none` early on indirect atom, so this is basically just a style concern, but using `u3_atom` instead of `c3_w` implies that the value might be an indirect atom.

Reply:
Changed to c3_w.